server/util/db/db.go

```diff
@@ -119,12 +121,80 @@ func maybeRunMigrations(dialect string, gdb *gorm.DB) error {
+	recordStartTime := func(db *gorm.DB) {
+		if db.DryRun || db.Statement == nil {
+			return
+		}
+		db.Statement.Settings.Store(gormStmtStartTimeKey, time.Now())
+	}
+	gdb.Callback().Create().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Delete().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Query().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Raw().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Row().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Update().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+
+	// Add callback that runs after other callbacks that records executed queries and their durations.
+	recordMetrics := func(db *gorm.DB) {
+		if db.DryRun || db.Statement == nil {
+			return
+		}
+		labels := prometheus.Labels{
+			metrics.SQLQueryTemplateLabel: db.Statement.SQL.String(),
+		}
+		v, _ := db.Statement.Settings.LoadAndDelete(gormStmtStartTimeKey)
+		if opStartTime, ok := v.(time.Time); ok {
+			metrics.SQLQueryDurationUsec.With(labels).Observe(float64(time.Now().Sub(opStartTime).Microseconds()))
+		}
+		// Ignore "record not found" errors as they don't generally indicate a
+		// problem with the server.
+		if db.Error != nil && !errors.Is(db.Error, gorm.ErrRecordNotFound) {
+			metrics.SQLErrorCount.With(labels).Inc()
+		}
+	}
+	gdb.Callback().Create().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+	gdb.Callback().Delete().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+	gdb.Callback().Query().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+	gdb.Callback().Raw().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+	gdb.Callback().Row().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+	gdb.Callback().Update().After("*").Register("buildbuddy:record_metrics", recordMetrics)
+}
+
+func openDB(configurator *config.Configurator, dialect string, connString string) (*gorm.DB, error) {
+	var dialector gorm.Dialector
+	switch dialect {
+	case "sqlite3":
+		dialector = sqlite.Open(connString)
+	case "mysql":
+		// Set default string size to 255 to avoid unnecessary schema modifications by GORM.
+		// Newer versions of GORM use a smaller default size (191) to account for InnoDB index limits
+		// that don't apply to modern MysQL installations.
+		dialector = mysql.New(mysql.Config{DSN: connString, DefaultStringSize: 255})
+	case "postgres":
+		dialector = postgres.Open(connString)
+	default:
+		return nil, fmt.Errorf("unsupported database dialect %s", dialect)
+	}
+
+	var l logger.Interface
+	l = sqlLogger{delegate: logger.Default, logLevel: logger.Warn}
+	if configurator.GetDatabaseConfig().LogQueries {
+		l = l.LogMode(logger.Info)
+	}
+	config := gorm.Config{
+		Logger: l,
+	}
+	gdb, err := gorm.Open(dialector, &config)
```

Comment:
"l" is nil if LogQueries is not defined. That might be OK in this case -- but you could you rephrase this in a way that is maybe safer by doing:

config := gorm.Config{}
if configurator.GetDatabaseConfig().LogQueries {
                 l := sqlLogger{delegate: logger.Default, logLevel: logger.Warn}
		l = l.LogMode(logger.Info)
                config.Logger = l
}

gdb, err := gorm.Open(dialector, &config)

Reply:
l is initialized immediately after l is defined so it should never be nil.